File: server/audit/tool_runner.py
"""Subprocess wrapper for external audit tools (semgrep, gitleaks, lizard, …).

Design goals:
- Never abort the audit if a tool is missing → return a ToolUsage(status="missing")
- Never hang → hard timeout per invocation
- Normalize stdout/stderr so analyzers can parse without special-casing
"""

from __future__ import annotations

import shutil
import subprocess
from dataclasses import dataclass
from pathlib import Path

from .schema import ToolUsage


@dataclass
class ToolResult:
    tool: str
    available: bool
    returncode: int = 0
    stdout: str = ""
    stderr: str = ""
    timed_out: bool = False
    version: str | None = None

    def to_usage(self, stack: str | None = None) -> ToolUsage:
        if not self.available:
            return ToolUsage(
                name=self.tool, status="missing", stack=stack,
                message="binary not found on PATH",
            )
        if self.timed_out:
            return ToolUsage(name=self.tool, status="timeout", stack=stack)
        if self.returncode not in (0, 1):  # many linters exit 1 on findings
            return ToolUsage(
                name=self.tool, status="error", stack=stack, version=self.version,
                message=f"exit {self.returncode}: {self.stderr[:200]}",
            )
        return ToolUsage(name=self.tool, status="ok", version=self.version, stack=stack)


def which(tool: str) -> str | None:
    return shutil.which(tool)


def run_tool(
    argv: list[str],
    *,
    cwd: Path | None = None,
    timeout: float = 120.0,
    env: dict[str, str] | None = None,
) -> ToolResult:
    """Execute `argv` with a timeout. Never raises; returns a ToolResult."""
    tool = argv[0]
    if which(tool) is None:
        return ToolResult(tool=tool, available=False)

    try:
        proc = subprocess.run(
            argv,
            cwd=str(cwd) if cwd else None,
            capture_output=True,
            text=True,
            timeout=timeout,
            env=env,
            check=False,
        )
        return ToolResult(
            tool=tool,
            available=True,
            returncode=proc.returncode,
            stdout=proc.stdout or "",
            stderr=proc.stderr or "",
        )
    except subprocess.TimeoutExpired:
        return ToolResult(tool=tool, available=True, timed_out=True, returncode=-1)
    except (OSError, ValueError) as exc:
        return ToolResult(
            tool=tool, available=True, returncode=-1, stderr=str(exc),
        )


def detect_version(tool: str, version_arg: str = "--version", timeout: float = 5.0) -> str | None:
    """Best-effort tool version detection — never raises."""
    if which(tool) is None:
        return None
    try:
        proc = subprocess.run(
            [tool, version_arg], capture_output=True, text=True, timeout=timeout, check=False,
        )
        raw = (proc.stdout or proc.stderr or "").strip().splitlines()
        return raw[0][:120] if raw else None
    except (subprocess.TimeoutExpired, OSError):
        return None

File: .quality/scripts/audit/lib/tool-runner.mjs
/**
 * lib/tool-runner.mjs — Subprocess wrapper for optional external audit tools (UC-663).
 *
 * Mirrors server/audit/tool_runner.py: every external tool (semgrep, gitleaks,
 * pip-audit, npm, lizard, jscpd, checkov) is OPTIONAL. A missing binary is
 * reported in tools_used with status "missing" and never aborts the audit.
 *
 * Uses spawnSync so a slow tool can be bounded by a timeout. We resolve the
 * binary with `which` first so "missing" is distinguishable from "ran and
 * failed" (spawnSync ENOENT is also caught as a fallback).
 */

import { spawnSync } from 'node:child_process';

/**
 * @typedef {Object} ToolResult
 * @property {boolean} available - binary found and spawned
 * @property {boolean} timedOut
 * @property {number|null} returncode
 * @property {string} stdout
 * @property {string} stderr
 */

/** True if `bin` resolves on PATH. */
export function isAvailable(bin) {
  const r = spawnSync('which', [bin], { encoding: 'utf-8' });
  return r.status === 0 && Boolean((r.stdout || '').trim());
}

/**
 * Run an external tool with a timeout. Never throws.
 *
 * @param {string[]} argv - [bin, ...args]
 * @param {object} [opts]
 * @param {string} [opts.cwd]
 * @param {number} [opts.timeout] - seconds (mirrors python float seconds)
 * @param {number} [opts.maxBuffer] - bytes (default 16 MiB for big JSON reports)
 * @returns {ToolResult}
 */
export function runTool(argv, opts = {}) {
  const [bin, ...args] = argv;
  const result = {
    available: false,
    timedOut: false,
    returncode: null,
    stdout: '',
    stderr: '',
  };
  if (!isAvailable(bin)) {
    return result;
  }
  let r;
  try {
    r = spawnSync(bin, args, {
      cwd: opts.cwd,
      encoding: 'utf-8',
      timeout: (opts.timeout ?? 120) * 1000,
      maxBuffer: opts.maxBuffer ?? 16 * 1024 * 1024,
    });
  } catch {
    // ENOENT despite `which` (race) or other spawn error → treat as missing.
    return result;
  }
  result.available = true;
  result.stdout = r.stdout || '';
  result.stderr = r.stderr || '';
  result.returncode = r.status;
  // spawnSync sets r.signal === 'SIGTERM' and r.error.code 'ETIMEDOUT' on timeout.
  if (r.error && (r.error.code === 'ETIMEDOUT' || r.signal === 'SIGTERM')) {
    result.timedOut = true;
  }
  return result;
}

/**
 * Best-effort version string for a tool (mirrors detect_version).
 * @param {string} bin
 * @param {string} [flag] - default "--version"
 * @returns {string|null}
 */
export function detectVersion(bin, flag = '--version') {
  if (!isAvailable(bin)) return null;
  const r = spawnSync(bin, [flag], { encoding: 'utf-8', timeout: 10_000 });
  const out = ((r.stdout || '') + (r.stderr || '')).trim();
  return out ? out.split('\n')[0].trim() : null;
}
